### PR TITLE
[5.2] Improve blade @endsection, overwrite by default.

### PR DESCRIPTION
**Pull request by jesseleite**
This addresses my concerns in issue #8970, regarding reusing and nesting `@extend @yield @section` directives within partials.

TL;DR...

``` php
@endsection // overwrite=false, does not allow multiple instances of section name.
```

``` php
@stop       // overwrite=false, does not allow multiple instances of section name.
```

``` php
@overwrite  // overwrite=true, allows multiple instances by overwriting at compile time.
```

I think most users would benefit from being able to use reuse and nest `@extend @yield @section` directives throughout an app, even if sections reuse common names like 'title' or 'body' in their partials.

We already have this functionality if we force the compiler to overwrite the section in memory by using `@overwrite` instead of `@endsection`.  However, default `@endsection` functionality disallows this, because it duplicates `@stop` functionality.  I think the issue here is in the naming of `@endsection`, and confusion around it's default behaviour.  Sure, using `@overwrite` solves #8970 functionally, but the terminology of 'overwrite' is misleading for this use case; I'm not trying to 'overwrite' the last title section, I'm trying to define a new title section in a totally separate partial.  I'm not doing the overwriting, the compiler is just overwriting the last instance at compile time.

Unfortunately, this IS a breaking change, even if it doesn't affect many users.  If merged, any users who need original functionality can `@stop` their sections.  However, I think allowing `@extend @yield @section` with `@endsection` everywhere throughout an app is much more intuitive, so I'm proposing this PR for 5.2.

Thank you for considering this.

### CODE REVIEW

**Comment by JerseyMilker** [comment]
My tests were passing here. @GrahamCampbell can we trigger Travis to rebuild?

**Comment by GrahamCampbell** [comment]
Done. :)

**Comment by JerseyMilker** [comment]
Thank you kind sir.

**Comment by taylorotwell** [comment]
Can you post more about your use case?

**Comment by taylorotwell** [comment]
Meaning, real code from your app to show what you are trying to accomplish.

**Comment by JerseyMilker** [comment]
Sure, will do.

**Comment by JerseyMilker** [comment]
@taylorotwell Okay here we go. I have [example app here](https://github.com/JerseyMilker/laravel/tree/blade-nest-extends-yield-section/resources/views).

Here's my view structure:

Tree

Specifically the issue is with those nested component partials.  Each of the three component partials @extend() that base component partial.

Components

Because each of those components use a common section name, 'body'... When I use the `@endsection` directive, all three components are outputting `<p>Component 1</p>` in my final view.  The reason is this: `@endsection` behaves like `@stop`, which prevents the 'body' section from being overwritten with new content in the compiler instance.

The issue is because I'm using a common 'body' section between three partials here.  If I end my 'body' sections with `@overwrite`, then each component renders output as I'd expect.  Therefore, I can get around this issue by closing out my sections with the `@overwrite` directive.  It doesn't feel as intuitive as `@endsection` though.  As a front end developer, I'm not trying to 'overwrite' anything in my front end.  The term 'overwrite' is referring to what's happening behind the scenes, overwriting the section content stored in the compiler instance.

So again, my proposition is to have `@endsection` duplicate `@overwrite` functionality, instead of duplicating `@stop` functionality as it currently does.  I feel like `@stop` to prevent the compiler from overwriting a section, is more of an edge case.  Would `@overwrite` would be more beneficial as a default?  For 5.2+, if you are worried about breaking changes?

Thanks for considering this btw.

**Comment by JerseyMilker** [comment]
@taylorotwell Okay here we go. I have [example app here](https://github.com/JerseyMilker/laravel/tree/blade-nest-extends-yield-section/resources/views).

Here's my view structure:

Tree

Specifically the issue is with those nested component partials.  Each of the three component partials @extend() that base component partial.

Components

Because each of those components use a common section name, 'body'... When I use the `@endsection` directive, all three components are outputting `<p>Component 1</p>` in my final view.

The issue is because I'm using a common 'body' section key between three partials here.  The reason is this: `@endsection` behaves like `@stop`, which prevents all succeeding sections with 'body' key from being overwritten in the compiler instance.  If I end my 'body' sections with `@overwrite`, then each component renders as I'd expect.  Therefore, I can get around this issue by closing out my sections with the `@overwrite` directive.

It doesn't feel as intuitive as `@endsection` though.  People are used to `@endif`, `@endforeach`, etc. directives.  As a front end developer, I'm not trying to 'overwrite' anything in my front end.  The term 'overwrite' is referring to what's happening behind the scenes, overwriting the section content stored in the compiler instance.  I feel like `@stop` functionality, could be desirable in edge cases where someone might want to yield identical section content multiple times per request, but `@overwrite` would be more useful for component driven design.

For example, maybe our components are displayed using Bootstrap Modals, where there's a lot of ugly Bootstrap boilerplate to deal with.  Put that boilerplate in a base partial, and extend it:
```php
@extends('components.modal.base')

@section('title', 'Add Comment')

@section('body')
	{!! BootForm::open()->action('/comment') !!}
		{!! BootForm::hidden('order_id')->value($serviceOrder->id) !!}
		{!! BootForm::textarea('Comment', 'comment')->rows(2)->hideLabel() !!}
		{!! BootForm::submit('Save Comment') !!}
	{!! BootForm::close() !!}
@endsection
```

So again, my proposition is to have `@endsection` default to `@overwrite` functionality, so that multiple components could be rendered off that base partial.  If the user needs to prevent compiler overwriting, they can still use the `@stop` directive.  Would this kind of functionality make as a default for the `@endsection` directive?  Maybe for 5.2+, if you are worried about breaking changes?

Sorry for the novel, and thanks for considering this.

**Comment by JerseyMilker** [comment]
@taylorotwell Specifically the issue is with component partials extending a base component.

Components

Each of the three component partials extend that base component partial.  Because these three components use a common section key, 'body'... When I use the `@endsection` directive, all three components are outputting `<p>Component 1</p>` in my final view.  The reason is this: `@endsection` behaves like `@stop`, which prevents all succeeding 'body' sections from being overwritten in the compiler instance.  If I end my 'body' sections with `@overwrite`, then each component partial renders as I'd expect.  Therefore, I can get around this issue by closing out my sections with the `@overwrite` directive.  More on this in a bit.

A more practical example, maybe I have multiple Bootstrap Modal components, where there's a lot of ugly Bootstrap boilerplate to deal with.  Put that boilerplate in a base partial, and extend it:
```php
@extends('components.modal.base')

@section('title', 'Add Comment')

@section('body')
	{!! BootForm::open()->action('/comment') !!}
		{!! BootForm::hidden('order_id')->value($serviceOrder->id) !!}
		{!! BootForm::textarea('Comment', 'comment')->rows(2)->hideLabel() !!}
		{!! BootForm::submit('Save Comment') !!}
	{!! BootForm::close() !!}
@endsection
```
But again, I'd have to end those 'body' sections with `@overwrite`.  It doesn't feel as intuitive as `@endsection`.  People are used to `@endif`, `@endforeach`, etc. directives.  As a front end developer, I'm not trying to 'overwrite' anything in my front end.  The term 'overwrite' is referring to what's happening behind the scenes, overwriting the section content stored in the compiler instance.  I feel like `@stop` functionality, could be desirable in edge cases where someone might want to yield an identical section content multiple times per request ...but `@overwrite` would be more useful for component driven design.

Sorry for the novel, and thanks for considering this :)

**Comment by taylorotwell** [comment]
Going to hold off on this.
